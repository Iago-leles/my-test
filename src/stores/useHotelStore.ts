import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Hotel } from '@/types'

import hotelData from '@/db/hotel.json'

import { HotelOrder } from '@/enums/hotels'

export const useHotelStore = defineStore('hotel', () => {
  const hotels = ref<Hotel[]>(hotelData.flatMap((data) => data.hotels) as Hotel[])
  const selectedHotel = ref<Hotel | null>(null)
  const isLoading = ref(false)
  const currentPage = ref(1)
  const itemsPerPage = 10
  const selectedPlace = ref<{ value: string; label: string } | null>(null)
  const orderBy = ref(HotelOrder.RECOMMENDED)
  const orderOptions = [HotelOrder.RECOMMENDED, HotelOrder.BEST_RATED]
  const filterName = ref('')

  const allHotels = computed(() => {
    return hotelData
  })

  function setFilterName(value: string) {
    filterName.value = value
  }

  const filteredHotels = computed(() => {
    let result = hotelData.flatMap((data) => data.hotels || []) as Hotel[]

    if (selectedPlace.value?.value) {
      result = allHotels.value
        .filter((places) => places.placeId === Number(selectedPlace.value?.value))
        .flatMap((places) => places.hotels) as Hotel[]
    }

    if (filterName.value.trim()) {
      const searchTerm = filterName.value.trim().toLowerCase()
      result = result.filter((hotel) => hotel.name.toLowerCase().includes(searchTerm))
    }

    const sortedOptions: Record<HotelOrder, Hotel[]> = {
      [HotelOrder.RECOMMENDED]: [...result].sort((a, b) => Number(a.price) - Number(b.price)),
      [HotelOrder.BEST_RATED]: [...result].sort((a, b) => Number(b.stars) - Number(a.stars)),
    }

    return sortedOptions[orderBy.value]
  })

  const displayedHotels = computed(() => {
    const endIndex = currentPage.value * itemsPerPage
    return filteredHotels.value.slice(0, endIndex)
  })

  function fetchNextPage(index: number, done: () => void) {
    setTimeout(() => {
      if (displayedHotels.value.length < filteredHotels.value.length) {
        currentPage.value++
        done()

        return
      }

      done()
    }, 500)
  }

  function setSelectedPlace(place: { value: string; label: string } | null) {
    selectedPlace.value = place || null
  }

  function setSelectedHotel(hotel: Hotel | null) {
    selectedHotel.value = hotel || null
  }

  return {
    hotels,
    isLoading,
    filterName,
    setFilterName,
    filteredHotels,
    displayedHotels,
    selectedPlace,
    orderOptions,
    orderBy,
    selectedHotel,
    fetchNextPage,
    setSelectedPlace,
    setSelectedHotel,
  }
})
