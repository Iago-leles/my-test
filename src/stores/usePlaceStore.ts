import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Place } from 'src/types'

import placesData from 'src/db/place.json'

export const usePlaceStore = defineStore('place', () => {
  const places = ref<Place[]>(placesData)
  const selectedPlace = ref<{ label: string; value: string } | null>(null)
  const hasSearched = ref(false)

  const formattedPlaces = computed(() => {
    return places.value.map((place) => ({
      label: `${place.name}, ${place.state.name}`,
      value: String(place.placeId),
    }))
  })

  const selectedPlaceData = computed(() => {
    return formattedPlaces.value.find((place) => place.value === selectedPlace.value?.value)
  })

  function setSelectedPlace(place: { label: string; value: string } | null) {
    selectedPlace.value = place
  }

  function setHasSearched(value: boolean) {
    hasSearched.value = value
  }

  function resetSearch() {
    hasSearched.value = false
  }

  return {
    places,
    selectedPlace,
    hasSearched,
    formattedPlaces,
    selectedPlaceData,
    setSelectedPlace,
    setHasSearched,
    resetSearch,
  }
})
