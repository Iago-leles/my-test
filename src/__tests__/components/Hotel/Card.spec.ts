import { mount, flushPromises } from '@vue/test-utils'
import { expect, describe, it, vi } from 'vitest'
import { createTestingPinia } from '@pinia/testing'
import { useHotelStore } from '@/stores/useHotelStore'

import type { Hotel } from '@/types'

import HotelCard from '@/components/Hotel/Card.vue'

describe('HotelCard', () => {
  it('should match snapshot', () => {
    const hotel = {
      id: 1,
      name: 'Hotel Test',
      address: { fullAddress: '123 Test St' },
      stars: 4.5,
      price: 100,
      amenities: [],
      images: [],
      hasRefundableRoom: true,
    }

    const wrapper = mount(HotelCard, {
      props: { hotel: hotel as unknown as Hotel },
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            fakeApp: true,
          }),
        ],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should set selected hotel on click', async () => {
    const hotel = {
      id: 1,
      name: 'Hotel Test',
      address: { fullAddress: '123 Test St' },
      stars: 4.5,
      price: 100,
      amenities: [],
      images: [],
      hasRefundableRoom: true,
    }

    const pinia = createTestingPinia({
      createSpy: vi.fn,
      stubActions: false,
    })

    const wrapper = mount(HotelCard, {
      props: { hotel: hotel as unknown as Hotel },
      global: {
        plugins: [pinia],
      },
    })

    await wrapper.find('[data-testid="hotel-card"]').trigger('click')

    await flushPromises()

    const hotelStore = useHotelStore(pinia)
    expect(hotelStore.selectedHotel).toEqual(hotel)
  })

  it('should render amenities', () => {
    const hotel = {
      id: 1,
      name: 'Hotel Test',
      address: { fullAddress: '123 Test St' },
      stars: 4.5,
      price: 100,
      amenities: [
        {
          key: 'WI_FI',
          label: 'Internet',
        },
        {
          key: 'RESTAURANT',
          label: 'Restaurante',
        },
        {
          key: 'ROOM_SERVICE',
          label: 'Servi√ßo de quarto',
        },
      ],
      images: [],
      hasRefundableRoom: true,
    }

    const wrapper = mount(HotelCard, {
      props: { hotel: hotel as unknown as Hotel },
    })

    expect(wrapper.findAll('[data-testid="hotel-amenity"]')).toHaveLength(3)
  })
})
