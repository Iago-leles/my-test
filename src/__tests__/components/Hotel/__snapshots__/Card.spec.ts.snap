// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelCard > should match snapshot 1`] = `
"<div data-v-470e659a="" class="q-card hotel-card q-mb-md no-shadow">
  <div data-v-470e659a="" class="row">
    <div data-v-470e659a="" class="col-4">
      <div data-v-813af4b2="" data-v-470e659a="" class="relative-position">
        <div data-v-813af4b2="" class="q-carousel q-panel-parent q-carousel--without-padding q-carousel--arrows-horizontal carousel-card" style="height: 300px;">
          <div class="q-carousel__slides-container"></div>
          <div class="q-carousel__control q-carousel__arrow q-carousel__prev-arrow q-carousel__prev-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_left</i></span></button></div>
          <div class="q-carousel__control q-carousel__arrow q-carousel__next-arrow q-carousel__next-arrow--horizontal absolute flex flex-center"><button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense" tabindex="0" type="button" regular="true"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon notranslate material-icons" aria-hidden="true" role="img">chevron_right</i></span></button></div>
        </div>
        <!--v-if-->
      </div>
    </div>
    <div data-v-470e659a="" class="col-8 q-pa-none q-h-100" data-testid="hotel-card">
      <div data-v-470e659a="" class="row justify-between content-wrapper">
        <div data-v-470e659a="" class="q-pa-md hotel-info" style="max-width: 64%;">
          <div data-v-470e659a="" class="text-h5 text-weight-medium text-grey-8 ellipsis">Hotel Test</div>
          <div data-v-470e659a="" class="text-subtitle1 text-grey-8 ellipsis">123 Test St</div>
          <div data-v-470e659a="" class="row q-gutter-sm items-center q-mt-md"><span data-v-470e659a="" class="row items-center q-mt-none"><div data-v-470e659a="" class="text-grey-6 q-mr-xs text-caption">4.5</div><div data-v-470e659a="" class="q-rating row inline items-center q-rating--non-editable text-amber" style="font-size: 1.2em;" role="radiogroup" aria-readonly="true"><div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 1"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-amber" aria-hidden="true" role="presentation">grade</i></div><div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 2"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-amber" aria-hidden="true" role="presentation">grade</i></div><div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 3"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-amber" aria-hidden="true" role="presentation">grade</i></div><div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 4"><i class="q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-amber" aria-hidden="true" role="presentation">grade</i></div><div class="q-rating__icon-container flex flex-center" role="radio" aria-checked="false" aria-label="grade 5"><i class="q-icon notranslate material-icons q-rating__icon text-amber" aria-hidden="true" role="presentation">grade</i></div></div></span>
            <!--v-if-->
            <!--v-if-->
          </div>
          <div data-v-470e659a="" class="q-badge flex inline items-center no-wrap q-badge--single-line bg-grey-7 q-px-md text-weight-medium q-mt-sm text-subtitle2" role="status">hotels.card.refundable</div>
        </div>
        <div data-v-470e659a="" class="text-left card-price text-grey-8">
          <div data-v-470e659a="">
            <div data-v-470e659a="" class="text-caption">hotels.card.from</div>
            <div data-v-470e659a="" class="text-h6 text-weight-bold"><span data-v-470e659a="" class="text-caption q-mb-none">hotels.card.currency</span><span data-v-470e659a="" class="text-h4 q-mb-none">100,00</span></div>
            <p data-v-470e659a="" class="text-caption q-mb-none">hotels.card.per_night</p>
            <div data-v-470e659a="" class="text-caption text-weight-medium">hotels.card.taxes_included</div>
          </div><button data-v-470e659a="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-blue text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase q-mt-xl q-py-sm q-px-xl" style="font-size: 20px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">hotels.card.select</span></span></button>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
